### ISSUE TITLE
change in number of matches reported by gather with default settings in 3.2.3

### ISSUE THREAD

**Issue by taylorreiter** [dib-lab/sourmash#976]
I recently changed my snakemake workflow to sourmash version 3.2.3. In doing so, I noticed that a gather run decreased in the number of species it reported for a specific signature from 41 to 38. Given the release notes and #942, I figured that this might be because of a change to the behavior of parameter `--threshold-bp`. I had a slack conversation with @ctb indicating that I might need to set `--threshold-bp 0` to return to the old behavior. When I did that, I got 270 matches. 

I'm not concerned with the difference in results for my specific pipeline; it's easy for me to accept the changes or to roll back my version to get the previous behavior. However, I would like to better understand how the gather cutoffs function, and how/why they changed in 3.2.3. I'm filing this issue to outline the change in behavior I observed with different versions and different parameters with the gather command. I've attached the signature that I'm running gather on here, and I provide links to the databases I'm using below. 

Signature: [at_least_5_studies_vita_vars.sig.txt](https://github.com/dib-lab/sourmash/files/4565447/at_least_5_studies_vita_vars.sig.txt)

Databases:
almeida: https://osf.io/5jyzr/download
pasolli: https://osf.io/3vebw/download
nayfach: https://osf.io/y3vwb/download
genbank-d2-k31.tar.gz: https://s3-us-west-2.amazonaws.com/sourmash-databases/2018-03-29/genbank-d2-k31.tar.gz

Sourmash 3.0.1, default `--threshold-bp`, 41 gather results, smallest match 50.0 kbp. 
Output: [at_least_5_studies_vita_vars_3.0.1.csv.txt](https://github.com/dib-lab/sourmash/files/4566022/at_least_5_studies_vita_vars_3.0.1.csv.txt)
```
sourmash gather -o at_least_5_studies_vita_vars.csv --scaled 2000 -k 31 at_least_5_studies_vita_vars.sig genbank-d2-k31.sbt.json almeida-mags-k31.sbt.json pasolli-mags-k31.sbt.json nayfach-k31.sbt.json
```

Sourmash 3.0.1, `--threshold-bp 0`, (TBD, will update) gather results
```
sourmash gather -o at_least_5_studies_vita_vars.csv --scaled 2000 --threshold-bp 0 -k 31 at_least_5_studies_vita_vars.sig genbank-d2-k31.sbt.json almeida-mags-k31.sbt.json pasolli-mags-k31.sbt.json nayfach-k31.sbt.json
```

Sourmash 3.2.3, default `--threshold-bp`, 38 gather results, smallest recovered match 50.0 kbp
Output: [at_least_5_studies_vita_vars_3.2.3.csv.txt](https://github.com/dib-lab/sourmash/files/4566025/at_least_5_studies_vita_vars_3.2.3.csv.txt)

```
sourmash gather -o at_least_5_studies_vita_vars.csv --scaled 2000 -k 31 at_least_5_studies_vita_vars.sig genbank-d2-k31.sbt.json almeida-mags-k31.sbt.json pasolli-mags-k31.sbt.json nayfach-k31.sbt.json
```

Sourmash 3.2.3, `--threshold-bp 0`, 270 gather results, smallest recovered match 2.0 kbp
```
sourmash gather -o at_least_5_studies_vita_vars.csv --scaled 2000 --threshold-bp 0 -k 31 at_least_5_studies_vita_vars.sig genbank-d2-k31.sbt.json almeida-mags-k31.sbt.json pasolli-mags-k31.sbt.json nayfach-k31.sbt.json
```

Looking at the difference between to two versions with the default `--threshold-bp`, the 3.2.3 version is missing the last three lines that appear in the 3.0.1 version.

**Comment by ctb** [comment]
# investigating gather changes

two places where behavior could have changed

## in sbt.py, function gather()

the threshold calculation code looks the same, but in the `for leaf in self.find(...)` loop, the condition for returning has changed from OLD:

```
similarity = query.minhash.contained_by(leaf_e, True)
if similarity > 0.0:
    results.append(...)
```
            
to NEW
```
containment = query.minhash.contained_by(leaf_mh, True)
if containment >= threshold:
    results.append(...)
```

So, in the old, if _any_ result is returned, we kept it. In the new, it must be above threshold.

## in search.py, function gather()


In the OLD code, the post-SBT-gather loop broke out of the loop on this condition:
```
intersect_bp = cmp_scaled * len(intersect_orig_query_mins)

if intersect_bp < threshold_bp:   # hard cutoff for now
    notify('found less than {} in common. => exiting',
           format_bp(threshold_bp))
    break
```

In the new code, this is done only when there are no matches from the individual database gather.

## what we see

we are seeing that the OLD behavior produced results between `--threshold-bp=5e4` and `--threshold-bp=0`.

My guess is that

1. sbt.py::gather was returning hits that were below threshold.
2. search.py::gather was letting by some hits that were below threshold.

(1) does appear to be happening in the code base, and I need to understand why... oh. That's because the search function (sbtmh.py::GatherMinHashes::search) is not using threshold at all!?

Will look into (2).

**Comment by ctb** [comment]
## more - 

in the old code, the calculation of `intersect_bp = cmp_scaled * len(intersect_orig_query_mins)` is at fault for being _less_ restrictive than the threshold_bp approach we are using now.

Briefly, we have two choices for overlap estimation. We can look at overlap estimation for the _original_ query or for the _remaining_ query. In the old code, we used the original query, which contained many more hashes and hence had a much larger overlap. In the new code, we are using the remaining query instead.

This is more of a decision than a bug. I'm not actually sure what the correct behavior is.

BUT:

This does explain the need to set threshold_bp=0 for these tests --

`test_gather_metagenome`,
`test_gather_metagenome_traverse`,
`test_gather_save_matches`

-- I verified this by looking at the cutoffs manually. For example, in `test_gather_metagenome`, when we look at the matches being returned, we see the `orig_intersect_bp` (2nd column) and the remnant `intersect_bp` (last column) - with threshold_bp=5e4, the last match/last row doesn't meet the threshold.

```
XXX 4870000 0.0 4870000
XXX 1920000 0.0 1920000
XXX 1880000 0.0 1690000
XXX 1570000 0.0 1570000
XXX 4850000 0.0 1520000
XXX 1900000 0.0 920000
XXX 4720000 0.0 760000
XXX 4690000 0.0 590000
XXX 4570000 0.0 420000
XXX 4270000 0.0 310000
XXX 4710000 0.0 70000
XXX 4450000 0.0 20000  # this last match is there in current behavior
```

This exact behavior does not explain `test_gather_metagenome_downsample`, directly; here I think the behavior seems to be because `scaled` is set to 100,000, while `--threshold-bp` is less than that.

I'm thinking about what to do there, too.

**Comment by ctb** [comment]
regarding

>Briefly, we have two choices for overlap estimation. We can look at overlap estimation for the original query or for the remaining query. In the old code, we used the original query, which contained many more hashes and hence had a much larger overlap. In the new code, we are using the remaining query instead.

I think, in the context of the gather algorithm, the current behavior is correct - when the overlap with the remaining query goes below (say) 50kb, we should stop reporting.  I will update docstring in #1001.

**Comment by ctb** [comment]
the `test_gather_metagenome_downsample` test is incorrectly passing; threshold_bp should be rounded down to a multiple of scaled, so as to be fully sensitive to all matches (within estimation ability). Implemented in #1001.

**Comment by ctb** [comment]
so, in sum, here we --

* uncovered a bug where matches < threshold were being returned from sbt gather, which also has performance impact;
* uncovered a bug in the way we were dealing with threshold_bp being a non-multiple of scaled;
* figured out that in #942 we switched behavior and meaning of `--threshold-bp` from old meaning (common hashes w/original query sig) to new meaning (common hashes w/remaining query sig).

The last point is up for debate as to what we should do, but the others are all bugs that we will fix in #1001.

**Comment by ctb** [comment]
Returning to the details in this issue --

* good matches were being incorrectly omitted when threshold_bp was not a multiple of scaled. This probably had minor impact.
* matches that were previously labeled as "good" are now "bad" based on how gather works underneath - the threshold-bp parameter _used_ to measure hashes shared with original query, but now measures hashes shared with remaining query.
* `--threshold-bp=0` will now return any and all matches with any overlap, which I think is appropriate. But this can lead to a lot of matches :)

If we want to keep this changed behavior as in #1001, my direction for people who are interested in retaining the old behavior would be to use `--threshold-bp=0` and then do post-filtering on the CSV to select the threshold based on the conditions you want to use.

**Comment by ctb** [comment]
Fixed in #1001
